server/player/src/packet_handlers/inventory.rs

```diff
@@ -426,13 +426,66 @@ fn handle_shift_click(
 }
 
 fn handle_number_key(
-    _game: &mut Game,
-    _world: &mut World,
-    _player: Entity,
-    _packet: ClickWindow,
-    _key: u8,
+    game: &mut Game,
+    world: &mut World,
+    player: Entity,
+    packet: ClickWindow,
+    key: u8,
 ) -> anyhow::Result<()> {
-    // TODO
+    let slot: usize = packet.slot as usize;
+    let window = world.get::<Window>(player);
+    let accessor = window.accessor(world)?;
+
+    // The slot index used to index the hotbar
+    let hotbar_slot = SlotIndex {
+        area: Area::Hotbar,
+        slot: (key - 1) as usize,
+    };
+
+    // Convert the slot index into the network slot
+    let hotbar_slot = window
+        .convert_slot(hotbar_slot, player)
+        .ok_or_else(|| anyhow::anyhow!("invalid slot index"))?;
+
+    // Perform the swap
+    if let Some(hotbar_slot_stack) = accessor.remove_item_at(hotbar_slot)? {
+        // Handles the case where there is an item in both target slots
+
+        let stack_under_cursor = accessor.remove_item_at(slot)?;
+
+        accessor.set_item_at(slot, hotbar_slot_stack)?;
+
+        if let Some(stack) = stack_under_cursor {
+            accessor.set_item_at(hotbar_slot, stack)?;
+        };
+    } else {
+        // Handles the case where there is an item in only the slot below the cursor
+
+        let stack_under_cursor = accessor.remove_item_at(slot)?;
+
+        if let Some(stack) = stack_under_cursor {
+            accessor.set_item_at(hotbar_slot, stack)?;
+        };
+    };
+
+    drop(accessor);
+
+    // Create a list of slots that were updated with this operation so they can be sent to handlers
+    let slots = smallvec![
+        window
+            .convert_network(slot as usize)
+            .ok_or_else(|| anyhow::anyhow!("invalid slot index"))?
+            .into(),
+        window
+            .convert_network((35 + key) as usize)
```

Comment:
These can also be listed as `SlotIndex { area: Area::Hotbar, key }`; there's no need to do an intermediate conversion.

Reply:
Thanks!